src/exchangeRateRouter.ts

```diff
@@ -301,6 +310,108 @@ const sendExchangeRates: express.RequestHandler = (req, res, next): void => {
   res.json({ data: exReq.requestedRatesResult.data })
 }
 
+const sendCoinranks: express.RequestHandler = async (
+  req,
+  res,
+  next
+): Promise<void> => {
+  const { ratesServerAddress } = config
+  const now = new Date()
+  const nowTimestamp = now.getTime()
+
+  const exReq = req as ExpressRequest
+  if (exReq == null) return next(500)
+
+  let query: CoinrankReq
+  try {
+    query = asCoinrankReq(req.query)
+  } catch (e) {
+    res
+      .status(400)
+      .send(`Error: ${e instanceof Error ? e.message : 'Unknown error'}`)
+    return
+  }
+  const { fiatCode, start, length } = query
+
+  try {
+    if (start < 1 || start > 2000) {
+      res
+        .status(400)
+        .send(`Invalid start param: ${start}. Must be between 1-2000`)
+      return
+    }
+    if (length < 1 || length > 100) {
+      res
+        .status(400)
+        .send(`Invalid length param: ${length}. Must be between 1-100`)
+      return
+    }
+    const fiatSuffix = fiatCode.split(':')[1]
+    const jsonString = await getAsync(
+      `${REDIS_COINRANK_KEY_PREFIX}_${fiatSuffix}`
+    )
+
+    let redisResult: CoinrankRedis = JSON.parse(jsonString)
+
+    if (fiatCode !== 'iso:USD') {
+      const lastUpdated =
+        redisResult != null ? new Date(redisResult.lastUpdate).getTime() : 0
+
+      // If no result in redis or it's out of date
+      if (redisResult == null || nowTimestamp - lastUpdated > EXPIRE_TIME) {
+        // Attempt to scale prices by foreign exchange rate
+
+        // Get exchange rate
+        const result = await fetch(
+          `${ratesServerAddress}/v2/exchangeRate?currency_pair=iso:USD_${fiatCode}`
+        )
+        const resultJson = await result.json()
+        const { exchangeRate } = asExchangeRateResponse(resultJson)
+        const rate = Number(exchangeRate)
+
+        // Get USD rankings
+        const jsonString = await getAsync(`${REDIS_COINRANK_KEY_PREFIX}_USD`)
+        redisResult = JSON.parse(jsonString)
+        const { markets } = redisResult
+        let data = markets.slice(start - 1, start + length)
+
+        // Modify the prices, mcap, & volume
+        data = data.map(m => ({
+          ...m,
+          price: m.price * rate,
+          marketCap: m.marketCap * rate,
+          volume24h: m.volume24h * rate
+        }))
+
+        res.json({ data })
+
+        // Update redis cache
+        const redisData: CoinrankRedis = {
+          markets: data,
```

Comment:
This only saves the the sliced data for the requested fiat. Convert everything in the USD list to the selected currency before saving it and returning to the user

Reply:
I don't understand. All USD market are queried by the engine. The API endpoint only writes non-USD markets after doing the conversion. The converted value is returned to the user and saved in redis.

Reply:
The data saved to redis is only a subset of the USD data (sliced at 378 and saved in line 392) and once it is saved subsequent requests for the same fiat code will no longer be null at line 361 which skips any conversion functionality.

Instead, take markets in line 377 and convert all of them to the requested fiat -> save to redis -> slice and return to user

Reply:
understood. makes sense.